### ISSUE TITLE
Ensure stable conda environments via pixi lockfile support

### ISSUE THREAD

**Issue by irm-codebase** [snakemake/snakemake#2880]
**Is your feature request related to a problem? Please describe.**
Conda is wonderful for scientific work, but one of the biggest headaches with it is ensuring that your environment stays stable as you develop software. Neither `conda` or `mamba` handle file locking very well, mostly because their environments are "centralized" (they live outside of projects and can be shared between them).

On top of that, you often need to constantly update your `environment.yaml` file to make sure things stay on point. This is often done manually, leading to headaches down the line. Tools like `poetry` can help, but they are often messy when it comes to interacting with `conda`.

**Describe the solution you'd like**
I recently discovered `pixi` thanks to this post: https://stackoverflow.com/questions/70851048/does-it-make-sense-to-use-conda-poetry

It's developed by the folks behind `mamba`, and I have to say: this looks like the next big thing when it comes to `conda` environment management:
- automatic lockfiles
- super fast
- environments are local to projects
- super well integrated with conda
- very easy/straightforward to use

Unfortunately... they do not support conversion to `environment.yaml`, and it seems like they won't for a while: https://github.com/prefix-dev/pixi/issues/800

It would be fantastic if `snakemake` got ahead in the game and implemented native support for this. I truly think it is a game changer in the `conda` world. I am picturing a lot of use cases for this, specially when turning workflows into modules that will be imported by other projects, or to break down very large workflows into subworflows/modules with their own environment/lockfile.

I am also very open to any workarounds that enable us to use both tools in tandem.

**Describe alternatives you've considered**
The aforementioned `poetry` + `conda` combination. And it was very underwhelming.

**Additional context**
Basically I want old environments to work without needing people to be pristine in how they handle them... and `pixi` solves that issue.

**Comment by sanjaynagi** [comment]
Totally agree @irm-codebase. Have a lot of snakemake workflows that use conda integration and as you say, it often causes headaches, particularly in larger workflows. 

From my use of it so far, Pixi does indeed seem like a great tool, so fast, getting faster and clearly well maintained. being able to specify lockfiles as well as/instead of conda envs would be ideal. 

I'd be happy to work on this, however, In the past I had a look at the conda script and it wasn't immediately obvious what would be necessary to implement pixi support. https://github.com/snakemake/snakemake/blob/main/snakemake/deployment/conda.py . If anyone is willing to provide some guidance, I am happy to look into this further and attempt to create a working draft.

**Comment by irm-codebase** [comment]
@sanjaynagi unfortunately I'm also struggling a bit here.

My approach to a design:
- Adding a command `pixilock` where you specify the location of the pixi lockfile.
- Adding  a command `pixienv` where you specify the name of the environment to use. This is then used to parse the pixi.lock file.

```python
rule download_basins_database:
    message: "Download database of hydro basins."
    params: url = config["data-sources"]["hydro-basins"]
    output: "data/automatic/raw-hydro-basins.zip"
    pixilock: "../../pixi.lock" 
    pixienv: "hydro"
    shell:
        "curl -sSLo {output} '{params.url}'"
```

This approach is fairly flexible in where the `pixi.lock` file is located. pixi always puts this at the top of a software project, but snakemake's design is agnostic to this, so I think it's better to specify the file location and the chosen environment separately.

Next is the question of how to run the environment... and for that I'm not really sure. pixi already generates those, but snakemake uses a separate hash system that puts all environments in a temporary `.snakemake` folder...

In theory, you could re-generate the environment using conda and the software version of each package as specified in  pixi.lock, but it seems like a waste...

Better to ask @johanneskoester directly on this one since it's a core design decision...

**Comment by johnne** [comment]
I also think it'd be great to get native support for pixi in snakemake. The strategy I've tried is to specify environments in a `pixi.toml` file for each conda environment file I have in a workflow. Then I pre-install all environments with `pixi install -a` which creates those environments under `.pixi/envs`, then I can point to those environment subdirs in my rules with the `conda:` directive. It seems to work except that I run into a very odd bug (#3192).

But native support for pixi and/or conda lockfiles (#2129) would be great!

**Comment by biopv** [comment]
Completely agree - this would be a very valuable addition to snakemake.

**Comment by brynpickering** [comment]
> Unfortunately... they do not support conversion to environment.yaml, and it seems like they won't for a while: https://github.com/prefix-dev/pixi/issues/800

In case it helps with implementation, they do now have an export to `environment.yml` option (https://github.com/prefix-dev/pixi/pull/2003).

**Comment by cademirch** [comment]
Closing in favor of #3915
